features/storage/system_storage/SystemStorage.cpp

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) 2018 ARM Limited. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ * Licensed under the Apache License, Version 2.0 (the License); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "BlockDevice.h"
+#include "FileSystem.h"
+#include "FATFileSystem.h"
+#include "LittleFileSystem.h"
+
+#if COMPONENT_SPIF
+#include "SPIFBlockDevice.h"
+#endif
+
+#if COMPONENT_DATAFLASH
+#include "DataFlashBlockDevice.h"
+#endif
+
+#if COMPONENT_SD
+#include "SDBlockDevice.h"
+#endif
+
+using namespace mbed;
+
+MBED_WEAK BlockDevice *BlockDevice::get_default_instance()
+{
+#if COMPONENT_SPIF
+
+    static SPIFBlockDevice default_bd(
+        MBED_CONF_SPIF_DRIVER_SPI_MOSI,
+        MBED_CONF_SPIF_DRIVER_SPI_MISO,
+        MBED_CONF_SPIF_DRIVER_SPI_CLK,
+        MBED_CONF_SPIF_DRIVER_SPI_CS,
+        MBED_CONF_SPIF_DRIVER_SPI_FREQ
+    );
+
+    return &default_bd;
+
+#elif COMPONENT_DATAFLASH
+
+    static DataFlashBlockDevice default_bd(
+        MBED_CONF_DATAFLASH_SPI_MOSI,
+        MBED_CONF_DATAFLASH_SPI_MISO,
+        MBED_CONF_DATAFLASH_SPI_CLK,
+        MBED_CONF_DATAFLASH_SPI_CS
+    );
+
+    return &default_bd;
+
+#elif COMPONENT_SD
+
+    static SDBlockDevice default_bd(
+        MBED_CONF_SD_SPI_MOSI,
+        MBED_CONF_SD_SPI_MISO,
+        MBED_CONF_SD_SPI_CLK,
+        MBED_CONF_SD_SPI_CS
+    );
+
+    return &default_bd;
+
+#else
+
+    return NULL;
+
+#endif
+
+}
+
+MBED_WEAK FileSystem *FileSystem::get_default_instance()
+{
+#if COMPONENT_SPIF || COMPONENT_DATAFLASH
+
+    static LittleFileSystem default_fs("fs", BlockDevice::get_default_instance());
+
+    return &default_fs;
+
+#elif COMPONENT_SD
+
+    static FATFileSystem default_fs("fs", BlockDevice::get_default_instance());
```

Comment:
Please make mount point configurable define for global constant. It will improve maintainability.

How about multipartition case? What should be default mount point for second partition?

Reply:
> Please make mount point configurable define for global constant.

This is a good idea 👍 

> How about multipartition case? What should be default mount point for second partition?

This solution only provides a single filesystem, we don't have a design in Mbed OS to provide more than one global object. @bridadan is working on a bigger solution to this but it requires more changes than in this PR.

Reply:
@JaniSuonpera 
This solution intends to simplify the OOB implementation, and therefore we believe, it is better to eliminate the option of redefining the mount point by configuration.

Reply:
NOTE! From Cloud Client point of view, we need to be able to support multipartitions and also still have LittleFS on the SD-card. Combinations such as 
1) SPI-flash with LittleFS
2) SPI-flash with FATFS
3) SPI-flash with 2 FAT FS partitions (working, backup)
4) SD-card with FATFS
5) SD-card with LittleFS
6) SD-card with multiple partitions (2 or more, working, backup)
must be possible.

Reply:
@JanneKiiskila As mentioned several times before: You have all the capabilities. We are talking about the OOB experience

Reply:
The moment you have multiple mount points in active use, splitting stuff between them, you're out-of-scope for this API. The "default" will just point to one of them - it should be whatever the most "general-purpose" working area is, so trying to put stuff in `/default/myapp/foo` is a reasonable thing for a simple app to attempt with no other information.

For a more complex system, startup configuration code should be setting up the multiple mount points under whatever names, and telling individual libraries explicitly what paths to use (eg `/sd/images/`), and hence which device they're on.

To aid interworking of these models we may want to add some alias support to `FileBase`, so that a filesystem can be referred to both explicitly as `/sd/` or `/default/`, then the `get_default_instance` could appear as both.

Reply:
Here's a proposal on that:

https://github.com/kjbracey-arm/mbed-os/commit/9273d4b95c824a82a71a1ca3ae4052cbfeaea93b

That would allow you to do 

     FileSystem *FileSystem::get_default_instance()
     {
         // create (or alternatively find) the sdcard, under "/sd"
         static FATFileSystem sdcard("sd", BlockDevice::get_default_instance()); 
         // make it respond to "/default" too
         sdcard.set_as_default(); 
         return &sdcard;
     }

Maybe in future that could be generalised to a full alias scheme - don't think it blocks such a generalisation.

Reply:
@yossi2le In this case we must override FileSystem::get_default_instance() under multipartition configuration.